### ISSUE TITLE
Severe Rendering Artifacts and Component Corruption in Bookcase Scene

### ISSUE THREAD

**Issue by RusticWolf86** [Kallyelish/Ref-Unlock#3]
Environment:

CUDA 11.8, PyTorch 2.0.1
NVIDIA A100 80GB
Docker container setup following provided instructions

Dataset:

COLMAP reconstructed bookcase scene (44 images)
Processed through complete pipeline: DSRNet + Depth Anything V2 + depth_params.json generation (using make_depth_scale.py)

Issue Description:
Despite following the complete preprocessing and training pipeline, the model produces severely corrupted outputs with color artifacts and component separation failure.
Training Results:

Completed 30,000 iterations without errors
Final metrics: L1 Loss ≈ 0.53, PSNR ≈ 9 (significantly below paper results)

Visual Results:
The attached images show:

1. Input preprocessing appears correct:

- Original images: Clear bookcase with glass reflections
- DSRNet clean images: Proper reflection removal
- Depth maps: Reasonable depth estimation

2. Output components severely corrupted:

- renders/: Color distortion with purple/green artifacts
- ref_color/ and trans_color/: Similar corruption patterns
- comp_ref_color/ and comp_trans_color/: Mostly black outputs
- ref_map/: Appears reasonable (grayscale reflection map)

Training Parameters Used:

- λIb = 0.8, SIbtrans = 10
- λinit = 0.01, λdepth = 30  
- λbi = 0.001, λref = 0.001
- SH degree = 5, iterations = 30000

Questions:

1. Are there known issues with indoor glass reflection scenes?
2. Could the component separation be failing due to specific DSRNet preprocessing characteristics?
3. Should certain hyperparameters be adjusted for this type of reflective scene?

Additional Context:
The same preprocessing pipeline was tested on multiple scenes with consistent corruption patterns, suggesting a systematic issue rather than dataset-specific problems.
Would appreciate guidance on debugging this issue or confirmation if this represents a known limitation.

The following are the respective input images - images, clean images and depth images

<img width="1871" height="1014" alt="Image" src="https://github.com/user-attachments/assets/cefcad4a-1d5f-4dfc-9c23-35c4c517965c" />

The following are the respective output images after rendering from render.py

<img width="1870" height="1016" alt="Image" src="https://github.com/user-attachments/assets/438c8ed8-06c9-42de-b652-56912da96c42" />

**Comment by Kallyelish** [comment]
Thank you for your interest in our work.

For **RFFR dataset reproduction**, the current code should work without any modifications.  

From your results, it seems that the reproduction is not working correctly. In particular, the visualizations of `trans_image (comp_trans_color)` and `render` look very different from the `clean_image` and `gt`.

In our normal reproduction, we could reach PSNR ≈ 27 at just 2000 iterations, and the expected visualizations of `ref_image` and `trans_image` look as shown in the figure below.

<img width="2370" height="1430" alt="Image" src="https://github.com/user-attachments/assets/eafd7260-51e9-4e39-87f3-68b99b966505" />
<img width="2425" height="1218" alt="Image" src="https://github.com/user-attachments/assets/b77f8f57-f280-4816-ac64-3af5adacbdcb" />

Since we have not encountered similar issues ourselves, based on your visualization I can only provide some rough and divergent debugging directions. You may try to check the following points:

1. **Check dataset path loading**  
   - Please print and verify whether the program is actually reading the correct paths for `images/` and `clean_images/`.  
   - A common issue is that the code does not read from the intended `images` and `clean_images`.  
   - Could you also share the exact `train` and `render` commands you used? (Note: the `-s` flag should point to the *root folder*, e.g., `bookcase/`).

2. **Check submodule installations**  
   - When setting up the environment, please confirm that you reinstalled all three submodules.  
   - For each submodule, you need to run:
     ```bash
     python setup.py install
     ```
   - If this step is skipped, environment mismatches can easily lead to abnormal results.

   - Our environment is:
     ```
     Python 3.9.19
     CUDA 11.7
     PyTorch 1.13.1+cu117
     ```

I would strongly recommend first verifying the dataset paths, and then checking if the issue persists.

If you still encounter any issues, please feel free to ask.

**Comment by RusticWolf86** [comment]
Thanks for the quick response. I’ll share the exact commands and checks I’ve tried so far.

## Training & Rendering Commands

For the bookcase scene I train with:
```
(ref_unlock_exact) root@bacc5eb10ec9:/workspace/Ref-Unlock/NeRF_Reflections/bookcase# \
python /workspace/Ref-Unlock/train.py \
    -s . \
    -d ./depth \
    --eval \
    -m /workspace/results/bookcase_author_1 \
    --iterations 2000 \
    2>&1 | tee /workspace/results/training_$(date +%Y%m%d_%H%M%S).log
```

And render with:
```
(ref_unlock_exact) root@bacc5eb10ec9:/workspace/Ref-Unlock/NeRF_Reflections/bookcase# \
python /workspace/Ref-Unlock/render.py \
    -m /workspace/results/bookcase_author_1/ \
    -s . \
    -d ./depth \
    --eval \
    --render_images \
    --skip_train
```

Scene folder structure at runtime:
```
bookcase/
├── images/
├── clean_images/
├── images_quarter/
├── clean_images_quarter/
├── depth/
├── distorted/
├── sparse/
├── stereo/
├── poses_bounds.npy
├── train.txt
└── val.txt
```
### Environment

Tried with CUDA 11.8 + PyTorch 2.0.1 (my main setup)

Also recreated the environment with CUDA 11.7 + PyTorch 1.13.1 (Python 3.9.19) as you suggested.
→ Unfortunately, no difference in results.

I also confirmed that all submodules were rebuilt with:
```
python setup.py install
```
### Additional Experiment (art3)

I also tried with another scene (art3) where I wrote my own code to generate the undistorted images. Training and rendering logs below:

### Training (2000 iters):
```
Number of points at initialisation : 5828
Training progress: 100%|...| 2000/2000 [00:42<00:00, 47.60it/s, Loss=0.907, Depth Loss=0.0002483]
```
```
[ITER 2000] Evaluating test:  L1 0.1510  PSNR 14.84
[ITER 2000] Evaluating train: L1 0.1489  PSNR 14.91
Training complete.
```

### Rendering:
```
Loading trained model at iteration 2000
Rendering progress: 100%|██████████████████████████████████████████████████████████| 6/6 [00:06<00:00, 1.13s/it]
Test FPS: 203.18
```

### Commands used:

```
(ref_unlock_exact) root@bacc5eb10ec9:/workspace/Ref-Unlock/NeRF_Reflections/art3/undistorted# \
python /workspace/Ref-Unlock/render.py \
    -m /workspace/results/art3_author_1 \
    -s . \
    -d ./depth \
    --eval \
    --render_images \
    --skip_train
```
```
(ref_unlock_exact) root@bacc5eb10ec9:/workspace/Ref-Unlock/NeRF_Reflections/art3/undistorted# \
python /workspace/Ref-Unlock/train.py \
    -s . \
    -d ./depth \
    --eval \
    -m /workspace/results/art3_author_1 \
    --iterations 2000 \
    2>&1 | tee /workspace/results/training_art3_author_1_$(date +%Y%m%d_%H%M%S).log
```
### Depth Parameters

For both scenes I generated depth_params.json using the provided ``` utils/make_depth_scale.py ```

### Documentation

To keep track, I’ve also created a detailed Markdown file documenting every step of my setup, preprocessing, and training procedure.

Despite these checks (paths, submodules, environment match, depth params), the corrupted outputs persist. Let me know if you’d like me to share the MD doc or a minimal repro package with scene + config.

I feel like i am making some mistake in the Depth anything V2 and RRM model - but while comparing to your images - i am getting the same images (using Depth Anything V2 - vitl encoder  and  DSRNet with dsrnet_l_4000_epoch33.pt weights).

**Comment by Kallyelish** [comment]
From your images, preprocessing steps, and commands, it seems like everything looks correct, and honestly I cannot directly see where the problem comes from. It might be best to rule things out step by step.

To fully exclude the possibility of path or command issues, I suggest:
- Try rewriting the command entirely in **absolute paths** and put it all in **one line** (maybe avoid using `\` for line breaks).  
- Add a print statement at [this line](https://github.com/Kallyelish/Ref-Unlock/blob/466431cfc2952810096b04c7cd4ff44b36a27ddd/scene/dataset_readers.py#L119) to double-check whether the program is indeed reading the correct `clean_images` path.

If that still does not solve it, on the **data side**, maybe try:
- Skipping COLMAP reconstruction and directly using the provided images and `sparse` folder from RFFR.  
- Running without `depth_params.json` to see if the issue persists.

I recommend experimenting with these variations to gradually narrow down where the bug might be.

**Comment by RusticWolf86** [comment]
I first tried running things with the complete paths and later switched to shorter paths.

1. Skipping COLMAP and directly using the provided images + sparse folder from RFFR ended up creating more issues than it solved. Using those images directly throws an error about needing undistorted images.

2.  When I use the ``` -d ``` argument, the code doesn’t run without a ``` depth_params.json ``` And without depth or clean images, the output quality isn’t really comparable — though I was still running into the same issues even when I included them.

That said, I really like the idea introduced in your paper. The way you make optimal use of higher-degree spherical harmonics is super clever. I’d definitely be interested in exploring this further, and I’m open to any potential collaboration if you see a fit. You can reach me at porus.vaid@gmail.com

In the meantime, if I figure out anything new I’ll share it with you, and if you come across any updates I’d love to hear from you as well.

**Comment by RusticWolf86** [comment]
Hi 👋,

Would it be possible for you to share the  **exact codebase**  you are currently running (for example, as a ZIP), so that I can reproduce the same results on my side?

It would also be really helpful if you could share the dataset you used with depth and clean images (even just a single subclass — e.g., bookcase — would be more than enough).

You can send it to me at [porus.vaid@gmail.com](mailto:porus.vaid@gmail.com)


Thanks a lot for your help!

**Comment by Kallyelish** [comment]
The code we are using is the same as the one available on GitHub. Maybe you can try using our preprocessed data: [https://drive.google.com/file/d/1QuUF86oOcoaDnsEjbbP9u9K-Nx3YJUfq/view?usp=sharing](https://drive.google.com/file/d/1QuUF86oOcoaDnsEjbbP9u9K-Nx3YJUfq/view?usp=sharing)

**Comment by RusticWolf86** [comment]
I created the environment from scratch and used the dataset you provided. I ran the training using your train.sh script with the appropriate paths. However, the issue still persists.

Observations

**Rendered from 3DGS (PostShot application):**

<img width="1918" height="1030" alt="Image" src="https://github.com/user-attachments/assets/0c553820-65a5-448b-bdcf-9e2bdaf65681" />

**Same viewpoint with gaussians after 30K iterations:**

<img width="1917" height="1018" alt="Image" src="https://github.com/user-attachments/assets/e9ec0fe1-ab42-4117-aff3-8db63c6ba453" />

**Entire gaussian scene generated by your code:**

<img width="1918" height="996" alt="Image" src="https://github.com/user-attachments/assets/749a38cd-1bcd-45b8-a81d-3115eda530fe" />

As you can see, the rendered outputs don’t align with what should be expected.

**Request**

If possible can you please share a zip file of your running code?
because we tired everything and somehow there is a major error that is not getting solved.

**Comment by Kallyelish** [comment]
Are the images rendered with `render.py` normal now?

**Comment by RusticWolf86** [comment]
No, they are coming the same like they were for bookcase.

Multiple artifacts and no proper view.

**Comment by Kallyelish** [comment]
Actually, after you raised this issue, we specifically pulled the code again from GitHub and tested it. The bookcase result shown above was produced using that GitHub code, which proves that the code itself is fine.

However, since you are indeed encountering issues on your side, that does seem a bit strange. Perhaps you could try pulling the code again, or double-check the command you’re using? Sorry, I also really can’t think of what else could be causing the problem.

**Comment by RusticWolf86** [comment]
Thank you for helping me with this issue :) I really appreciate your time and patience.

I've tried reproducing your results on multiple systems but keep running into the same rendering problems. Your approach to spherical harmonics and reflection disentanglement is really interesting, and it's got me thinking more about lighting in 3D reconstruction.

Would you be open to an online meet sometime? I feel like there might be something small I'm missing that would be easier to spot if you could see my setup directly. I'm flexible with timing and happy to work around your schedule.

I'm also excited about the possibility of future collaboration! Whether we solve this current issue or explore new directions in lighting-aware reconstruction, I'd love to contribute to this field under your guidance.

Thank you again for your openness to help - it means a lot to a researcher trying to break into this space!

**Comment by while0l1** [comment]
Replace the glm in `submodules/diff-gaussian-rasterization/third_party` with `glm-0.9.9.8` and it works for me

**Comment by wangxl0385** [comment]
> Replace the glm in `submodules/diff-gaussian-rasterization/third_party` with `glm-0.9.9.8` and it works for me

This works for me, too!
